tests/std/tests/GH_004597_self_swap/test.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+// Test GH-4597 "<utility>: Side effects in self-swaps of pair are skipped"
+struct swap_counter {
+    unsigned int* pcnt_ = nullptr;
+
+    friend constexpr void swap(swap_counter& lhs, swap_counter& rhs) noexcept {
+        std::swap(lhs.pcnt_, rhs.pcnt_);
+        if (lhs.pcnt_ != nullptr) {
+            ++(*lhs.pcnt_);
+        }
+        if (rhs.pcnt_ != nullptr) {
+            ++(*rhs.pcnt_);
+        }
+    }
+
+    constexpr bool operator==(unsigned int x) const {
+        return *pcnt_ == x;
+    }
+};
+
+
+void test_gh_4595() {
+    {
+        static_assert([] {
+            unsigned int cnt{};
+            std::pair<swap_counter, int> pr{swap_counter{&cnt}, 0};
+            pr.swap(pr);
+            return (cnt == 2u) && (pr.first == 2u) && (pr.second == 0);
+        }());

```

Comment:
Lambda-expression in constant expression is only valid in C++17. I think we can add some usual functions marked with `CONSTEXPR20`, and then always `assert` their results and `static_assert` their results since C++20.

... like this
```C++
#if _HAS_CXX20
static_assert(fun(args));
#endif // _HAS_CXX20
// ...
assert(fun(args));
```

Reply:
Done